File: paykit-ffi/bindings/android/lib/src/main/kotlin-manual/com/synonym/paykit/PaykitPublicKeys.kt
package com.synonym.paykit

public object PaykitPublicKeys {
    private const val PREFIX: String = "pubky"
    private const val RAW_LENGTH: Int = 52
    private val allowedCharacters: Set<Char> =
        "ybndrfg8ejkmcpqxot1uwisza345h769".toSet()

    @JvmStatic
    public fun normalize(value: String): String = "$PREFIX${raw(value)}"

    @JvmStatic
    public fun raw(value: String): String {
        val trimmed = value.trim()
        val rawValue = if (
            trimmed.startsWith(PREFIX) &&
            trimmed.length == PREFIX.length + RAW_LENGTH
        ) {
            trimmed.substring(PREFIX.length)
        } else {
            trimmed
        }
        require(rawValue.length == RAW_LENGTH && rawValue.all { it in allowedCharacters }) {
            "invalid Pubky public key"
        }
        return rawValue
    }

    @JvmStatic
    public fun redacted(value: String): String {
        val normalized = normalize(value)
        return "${normalized.take(PREFIX.length + 6)}...${normalized.takeLast(6)}"
    }
}

File: paykit-ffi/bindings/ios/PaykitPublicKeys.swift
import Foundation

public enum PaykitPublicKeyFormatError: Error, Equatable {
    case invalidPubkyPublicKey
}

public enum PaykitPublicKeys {
    private static let prefix = "pubky"
    private static let rawLength = 52
    private static let allowedCharacters = Set("ybndrfg8ejkmcpqxot1uwisza345h769")

    public static func normalize(_ value: String) throws -> String {
        "\(prefix)\(try raw(value))"
    }

    public static func raw(_ value: String) throws -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let rawValue: String
        if trimmed.hasPrefix(prefix), trimmed.count == prefix.count + rawLength {
            rawValue = String(trimmed.dropFirst(prefix.count))
        } else {
            rawValue = trimmed
        }
        guard rawValue.count == rawLength,
              rawValue.allSatisfy({ allowedCharacters.contains($0) }) else {
            throw PaykitPublicKeyFormatError.invalidPubkyPublicKey
        }
        return rawValue
    }

    public static func redacted(_ value: String) throws -> String {
        let normalized = try normalize(value)
        return "\(normalized.prefix(prefix.count + 6))...\(normalized.suffix(6))"
    }
}
